torchmetrics/functional/classification/auc.py

```diff
@@ -31,16 +31,16 @@ def _auc_update(x: Tensor, y: Tensor) -> Tuple[Tensor, Tensor]:
     return x, y
 
 
-def _auc_compute(x: Tensor, y: Tensor, reorder: bool = False) -> Tensor:
+def _auc_compute(x: Tensor, y: Tensor, reorder: bool = False, tol: float = 1e-6) -> Tensor:
     with torch.no_grad():
         if reorder:
             # TODO: include stable=True arg when pytorch v1.9 is released
             x, x_idx = torch.sort(x)
             y = y[x_idx]
 
         dx = x[1:] - x[:-1]
-        if (dx < 0).any():
-            if (dx <= 0).all():
+        if (dx + tol < 0).any():
+            if (dx <= tol).all():
```

Comment:
am I reading it correctly that if I have 

dx = [-1e-7, 5e-7, 5e-7, ... 10000 times ...., 5e-7] then we'll deduce incorrect direction here?

Reply:
yes, it seems that direction/sign is changed, so shall ve preserve it...

Reply:
maybe I am misunderstanding the question...
the tensor
```dx = [1e-7, 5e-7, 5e-7, ... 10000 times ...., 5e-7]```
implies that `direction=1` right? so if we say that numerical instability leads to a change of sign in the first element
```dx = [-1e-7, 5e-7, 5e-7, ... 10000 times ...., 5e-7]```
then it will still be `direction=1` with the change.

Reply:
Sorry, let's maybe use this instead:
`dx = [-1.0, 5e-7, 5e-7, ... 10000 times ...., 5e-7]`
Now both `(dx+tol<0).any()` and `(dx <= tol).all()` are true and we'll discover direction as -1.

Reply:
thanks, I see the problem now...
Do you have anyway to solve this?

Reply:
Numerical issues are usually tricky. I'm suggesting we remove checks for the internal callers of this function, let me quickly show an example.

Reply:
@SkafteNicki , check out https://github.com/PyTorchLightning/metrics/pull/230

Reply:
@maximsch2 looks good, closing this in favour of yours